core/src/main/scala/filodb.core/memstore/PartKeyLuceneIndex.scala

```diff
@@ -202,46 +212,54 @@ class PartKeyLuceneIndex(dataset: Dataset,
     // FIXME this API returns duplicate values because same value can be present in multiple lucene segments
-    val searcher = searcherManager.acquire()
-    val indexReader = searcher.getIndexReader
-    val segments = indexReader.leaves()
+
     val freqOrder = Ordering.by[TermInfo, Int](_.freq)
     val topkResults = new PriorityQueue[TermInfo](topK, freqOrder)
-    var termsRead = 0
-    segments.asScala.foreach { segment =>
-      val terms = segment.reader().terms(fieldName)
-
-      //scalastyle:off
-      if (terms != null) {
-        val termsEnum = terms.iterator()
-        var nextVal: BytesRef = termsEnum.next()
-        while (nextVal != null && termsRead < MAX_TERMS_TO_ITERATE) {
-          //scalastyle:on
-          val valu = BytesRef.deepCopyOf(nextVal) // copy is needed since lucene uses a mutable cursor underneath
-          val ret = new UTF8Str(valu.bytes, bytesRefToUnsafeOffset(valu.offset), valu.length)
-          val freq = termsEnum.docFreq()
-          if (topkResults.size < topK) {
-            topkResults.add(TermInfo(ret, freq))
-          }
-          else if (topkResults.peek.freq < freq) {
-            topkResults.remove()
-            topkResults.add(TermInfo(ret, freq))
+
+    withNewSearcher { searcher =>
+      val indexReader = searcher.getIndexReader
+      val segments = indexReader.leaves()
+      var termsRead = 0
+      segments.asScala.foreach { segment =>
+        val terms = segment.reader().terms(fieldName)
+
+        //scalastyle:off
+        if (terms != null) {
+          val termsEnum = terms.iterator()
+          var nextVal: BytesRef = termsEnum.next()
+          while (nextVal != null && termsRead < MAX_TERMS_TO_ITERATE) {
+            //scalastyle:on
+            val valu = BytesRef.deepCopyOf(nextVal) // copy is needed since lucene uses a mutable cursor underneath
+            val ret = new UTF8Str(valu.bytes, bytesRefToUnsafeOffset(valu.offset), valu.length)
+            val freq = termsEnum.docFreq()
+            if (topkResults.size < topK) {
+              topkResults.add(TermInfo(ret, freq))
+            }
+            else if (topkResults.peek.freq < freq) {
+              topkResults.remove()
+              topkResults.add(TermInfo(ret, freq))
+            }
+            nextVal = termsEnum.next()
+            termsRead += 1
           }
-          nextVal = termsEnum.next()
-          termsRead += 1
         }
       }
     }
     topkResults.toArray(new Array[TermInfo](0)).sortBy(-_.freq).toSeq
   }
 
-  def indexNames: scala.Iterator[String] = {
-    val searcher = searcherManager.acquire()
-    val indexReader = searcher.getIndexReader
-    val segments = indexReader.leaves()
-    segments.asScala.iterator.flatMap { segment =>
-      segment.reader().getFieldInfos.asScala.toIterator.map(_.name)
-    }.filterNot { n => ignoreIndexNames.contains(n) }
+  def indexNames(limit: Int): Seq[String] = {
+
+    var ret: Seq[String] = Nil
```

Comment:
See suggestion above so that `withNewSearcher` can return a value.  Makes the var totally unnecessary